macro/src/expand.rs

```diff
@@ -552,59 +552,7 @@ fn expand_cxx_function_shim(efn: &ExternFn, types: &Types) -> TokenStream {
-                                true => return quote_spanned!(span=> &::cxx::private::RustOption::from_option_string_mut(#var.assume_init()) as _),
-                            },
-                            Type::RustVec(vec) if vec.inner == RustString => match r.mutable {
-                                false => return quote_spanned!(span=> &::cxx::private::RustOption::from_option_vec_string_ref(#var.assume_init()) as _),
-                                true => return quote_spanned!(span=> &::cxx::private::RustOption::from_option_vec_string_mut(#var.assume_init()) as _),
-                            },
-                            Type::RustVec(_) => match r.mutable {
-                                false => return quote_spanned!(span=> &::cxx::private::RustOption::from_option_vec_ref(#var.assume_init()) as _),
-                                true => return quote_spanned!(span=> &::cxx::private::RustOption::from_option_vec_mut(#var.assume_init()) as _),
-                            },
-                            _ => {},
-                        }
-                        if types.is_considered_improper_ctype(&ty.inner) {
-                            if r.mutable {
-                                if r.pinned {
-                                    quote_spanned!(span=> &::cxx::RustOption::from_option_mut_improper_pinned(#var.assume_init()) as _)
-                                } else {
-                                    quote_spanned!(span=> &::cxx::RustOption::from_option_mut_improper(#var.assume_init()) as _)
-                                }
-                            } else if r.pinned {
-                                quote_spanned!(span=> &::cxx::RustOption::from_option_ref_improper_pinned(#var.assume_init()) as _)
-                            } else {
-                                quote_spanned!(span=> &::cxx::RustOption::from_option_ref_improper(#var.assume_init()) as _)
-                            }
-                        } else if r.mutable {
-                            if r.pinned {
-                                quote_spanned!(span=> &::cxx::RustOption::from_option_mut_pinned(#var.assume_init()) as _)
-                            } else {
-                                quote_spanned!(span=> &::cxx::RustOption::from_option_mut(#var.assume_init()) as _)
-                            }
-                        } else if r.pinned {
-                            quote_spanned!(span=> &::cxx::RustOption::from_option_ref_pinned(#var.assume_init()) as _)
-                        } else {
-                            quote_spanned!(span=> &::cxx::RustOption::from_option_ref(#var.assume_init()) as _)
-                        }
-                    }
-                    _ => unreachable!(),
-                }
-            }
+            Type::RustOption(_) => quote_spanned!(span=> #var.as_mut_ptr() as *const ::cxx::private::RustOption<_>),
```

Comment:
indirect ABI is way simpler :)

One thing I don't understand here though is why am I using *const instead of *mut. I am blindly copying the Type::RustVec case just above. I assume it doesn't matter because it probably just gets silently cast to a mut pointer across ffi (since the argument on the C++ side is not const) but I wonder if something subtle is happening?